packages/blockchain/src/consensus/base.ts

```diff
@@ -0,0 +1,72 @@
+// eslint-disable-next-line implicit-dependencies/no-implicit
+import type { LevelUp } from 'levelup'
+import { Block, BlockHeader } from '@ethereumjs/block'
+import Common from '@ethereumjs/common'
+import { DBManager } from '../db/manager'
+
+/**
+ * Base interface that a consensus class should implement.
+ */
+export interface ConsensusBase {
+  /**
+   * Initialize genesis for consensus mechanism
+   * @param genesisBlock genesis block
+   */
+  genesisInit?(genesisBlock: Block): Promise<void>
+
+  /**
+   * Set up consensus mechanism
+   */
+  setup?(): Promise<void>
+
+  /**
+   * Validate block
+   * @param block block to be validated
+   */
+  validate?(block: Block): Promise<void>
+
+  /**
+   * Update consensus on new block
+   * @param block new block
+   * @param commonAncestor common ancestor block header
+   * @param ancientHeaders array of ancestor block headers
+   */
+  newBlock?(block: Block, commonAncestor: BlockHeader, ancientHeaders: BlockHeader[]): Promise<void>
+
+  // eslint-disable-next-line
+  new(db: LevelUp, dbManager?: DBManager, _common?: Common, _validateConsensus?: boolean): ConsensusBase
+}
+
+/**
+ * Base class that a consensus class should extend.
+ */
+export class ConsensusBase implements ConsensusBase {
```

Comment:
I tried that at first, but problem is it breaks this case here in `index.ts` https://github.com/ethereumjs/ethereumjs-monorepo/pull/1823/files#diff-db407020ca647b02014b648ca3a09fa88c5278fae3055e19ffec073c33e96678R217. After going down a stackoverflow rabbit hole, the only technique I could find that typescript would accept was this way of merging a same-named interface and class, but the class couldn't be abstract for the technique to work. That was my understanding anyway - definitely not an expert on the finer points of typescript.

Reply:
your link isn't working for me. is the problem that they are the same name? you can just call the interface `Consensus`

Reply:
There shouldn't be any type of base class here, consensus implementations will just be too different, so no need to unify any code here, this will not lead to much.

Instead just do the interface and implement the interface for Clique and Ethash.

Reply:
Removed the class